partiql-logical-planner/src/typer.rs

```diff
@@ -0,0 +1,917 @@
+                ],
+            ),
+            vec![
+                StructField::new("id", int!()),
+                StructField::new("name", str!()),
+                StructField::new("age", int!()),
+            ],
+        )
+        .expect("Type");
+
+        // Open Schema with `Strict` typing mode and `bar` in nested attribute.
+        assert_query_typing(
+            TypingMode::Strict,
+            "SELECT customers.id, customers.name, customers.details.age, customers.details.foo.bar FROM customers",
+            create_customer_schema(true,vec![
+                StructField::new("id", int!()),
+                StructField::new("name", str!()),
+                StructField::new("details", details.clone()),
+            ]),
+            vec![
+                StructField::new("id", int!()),
+                StructField::new("name", str!()),
+                StructField::new("age", int!()),
+                StructField::new("bar", any!()),
+            ],
+        )
+        .expect("Type");
+    }
+
+    #[test]
+    fn simple_sfw_with_alias() {
+        // Open Schema with `Strict` typing mode and `age` in nested attribute.
+        let details_fields = struct_fields![("age", int!())];
+        let details = r#struct![BTreeSet::from([details_fields])];
+
+        assert_query_typing(
+            TypingMode::Strict,
+            "SELECT d.age FROM customers.details AS d",
+            create_customer_schema(
+                true,
+                vec![
+                    StructField::new("id", int!()),
+                    StructField::new("name", str!()),
+                    StructField::new("details", details.clone()),
+                ],
+            ),
+            vec![StructField::new("age", int!())],
+        )
+        .expect("Type");
+
+        assert_query_typing(
+            TypingMode::Strict,
+            "SELECT c.id AS my_id, customers.name AS my_name FROM customers AS c",
+            create_customer_schema(
+                false,
+                vec![
+                    StructField::new("id", int!()),
+                    StructField::new("name", str!()),
+                    StructField::new("age", any!()),
+                ],
+            ),
+            vec![
+                StructField::new("my_id", int!()),
+                StructField::new("my_name", str!()),
+            ],
+        )
+        .expect("Type");
+    }
+
+    #[test]
+    fn simple_sfw_err() {
+        // Closed Schema with `Strict` typing mode and `age` non-existent projection.
+        let err1 = r#"No Typing Information for SymbolPrimitive { value: "age", case: CaseInsensitive } in closed Schema PartiqlType(Struct(StructType { constraints: {Open(false), Fields([StructField { name: "id", ty: PartiqlType(Int) }, StructField { name: "name", ty: PartiqlType(String) }])} }))"#;
+        let err2 = r#"Unexpected Typing Environment; expected typing environment with only one type but found 2 types"#;
```

Comment:
May be resolved in an upcoming PR, but I feel like this second error is somewhat confusing. I was mostly confused by the wording of the error when it states it found 2 types in the typing environment when 1 was expected. Shouldn't there only be the first error, which indicates that `age` cannot be found in the typing environment?

It's classified as an `IllegalState` error which I thought was not due to any user input error.

Reply:
Let's discuss more as I still believe we need to have the list of all the errors we have encountered during typing (like an StackTrace). It's an `IllegalState` b/c we were expecting one type in the environment at some point during typing but found one. I can make changes in subsequent PRs if we agreed to.